src/Razor/src/Microsoft.VisualStudio.LanguageServerClient.Razor/HtmlCSharp/CompletionHandler.cs

```diff
@@ -262,23 +269,91 @@ static bool TryConvertToCompletionList(SumType<CompletionItem[], CompletionList>
         //    This isn't an applicable scenario for C# to provide the "for" keyword; however, Razor still wants that to be applicable because if you type out the full @for keyword it will generate the
         //    appropriate C# code.
         // 3. Remove Razor intrinsic design time items. Razor adds all sorts of C# helpers like __o, __builder etc. to aid in C# compilation/understanding; however, these aren't useful in regards to C# completion.
-        private static CompletionList PostProcessCSharpCompletionList(
+        private CompletionList PostProcessCSharpCompletionList(
             CompletionParams request,
             LSPDocumentSnapshot documentSnapshot,
             TextExtent? wordExtent,
             CompletionList completionList)
         {
+            var formattingOptions = _formattingOptionsProvider.GetOptions(documentSnapshot.Uri);
             if (IsSimpleImplicitExpression(request, documentSnapshot, wordExtent))
             {
                 completionList = RemovePreselection(completionList);
                 completionList = IncludeCSharpKeywords(completionList);
+
+                // -1 is to account for the transition so base indentation is "|@if" instead of "@|if"
+                var baseIndentation = Math.Max(GetBaseIndentation(wordExtent) - 1, 0);
+                completionList = IncludeCSharpSnippets(baseIndentation, completionList, formattingOptions);
+            }
+            else if (IsWordOnEmptyLine(wordExtent, documentSnapshot))
+            {
+                var baseIndentation = GetBaseIndentation(wordExtent);
+                completionList = IncludeCSharpSnippets(baseIndentation, completionList, formattingOptions);
             }
 
             completionList = RemoveDesignTimeItems(documentSnapshot, wordExtent, completionList);
 
             return completionList;
         }
 
+        private static CompletionList IncludeCSharpSnippets(int baseIndentation, CompletionList completionList, FormattingOptions formattingOptions)
+        {
+            var baseIndentationString = GetIndentationString(baseIndentation, formattingOptions);
+            var baseIndentationPlus1String = GetIndentationString(baseIndentation + formattingOptions.TabSize, formattingOptions);
+
+            var forSnippet = new CompletionItem()
```

Comment:
@dibarbet all these C# snippets should be included on the C# side, there's a possibility where we build out custom snippets for C# scenarios that are contextual. Meaning we can be smarter about where we add these snippets where pre-LSP we couldn't. 

I added them in Razor for now since it was faster but is there something on the C# side tracking additions like this?

Reply:
We do have this issue - https://github.com/dotnet/roslyn/issues/53853

These snippets should already exist right?  Is it not possible to convert the VS snippets to LSP ones?  I think the C# completion handler should be able to retrieve snippets as well - https://sourceroslyn.io/#Microsoft.CodeAnalysis.CSharp.Features/Completion/CompletionProviders/SnippetCompletionProvider.cs,30

We're just filtering them out right now here - https://sourceroslyn.io/#Microsoft.CodeAnalysis.LanguageServer.Protocol/Handler/Completion/CompletionHandler.cs,426

Reply:
> These snippets should already exist right?

They do but the pre-existing snippets are contextless. Meaning you'll even be fed the `prop` snippet in a method block lol.
![image](https://user-images.githubusercontent.com/2008729/125705078-c8fb5a29-66a5-4c76-9f65-1dccdee0790c.png). They're also expected to be invoked via tab+tab. For instance `if`, `for` etc. is a built-in snippet. Considering adding a new LSP method for `tab` next to a word identifier but that'd be for another time

There's a separate workitem to convert from built-in VS Snippet -> LSP snippet / even allow built-in snippets in LSP experiences [here](https://github.com/dotnet/aspnetcore/issues/30774).

Reply:
> They do but the pre-existing snippets are contextless. Meaning you'll even be fed the prop snippet in a method block lol.

In general right?  Perhaps we should just improve that on the c# side?  Or maybe just filter them out on the razor side

> They're also expected to be invoked via tab+tab. For instance if, for etc. is a built-in snippet. Considering adding a new LSP method for tab next to a word identifier but that'd be for another time

So we'd have to convert the roslyn snippets returned into LSP snippets, and in that case would that also not resolve the tab tab behavior since they'd be actual items in the list?  The tab tab behavior iirc is completely outside the LSP layer.

Reply:
> In general right? Perhaps we should just improve that on the c# side? Or maybe just filter them out on the razor side

Definitely something that could be improved; however, the difficult part with snippets is that they could be anything. Meaning a user could have modified them and on the C# side you don't entirely know what contexts the snippet is applicable in. 

> So we'd have to convert the roslyn snippets returned into LSP snippets, and in that case would that also not resolve the tab tab behavior since they'd be actual items in the list? The tab tab behavior iirc is completely outside the LSP layer.

Correct, you wouldn't have the tab + tab behavior. I've talked with Mika and friends about tab + tab and that'd have to be a separate effort where we add that concept into the LSP platform.

Honestly my take away after playing with the snippet system is that there's huge opportunity to improve it and there's a good amount of pain in how things are currently implemented. I'm not saying the LSP snippet system is superior but I do think it gives an opportunity for us to re-think how these experiences should playout for users, for the better.

Reply:
>Meaning a user could have modified them and on the C# side you don't entirely know what contexts the snippet is applicable in.

Very good point!

Regardless of the UI experience (which we can figure out later, I'm OK with this approach though) : 
I guess my overall point is that if we're supporting c# snippets they should come from the c# server, and we should have a method to just use the snippets produced by the roslyn completion service in the C# LSP handler.

I'm just thinking of the extra work we'll have to do once we support snippets in the c# lsp completion handler to remove these, and then filter out in razor based on context the snippets you want (otherwise these hardcoded ones will be duplicates).  And if it is possible to just do it one way once

Reply:
> I guess my overall point is that if we're supporting c# snippets they should come from the c# server, and we should have a method to just use the snippets produced by the roslyn completion service in the C# LSP handler.

Yup 100% agree, I just have this PR because I'm going on vaca in two days and it was the fastest way to get it done 😄 



> I'm just thinking of the extra work we'll have to do once we support snippets in the c# lsp completion handler to remove these

Ehhh we'll revert the commit, hopefully shouldn't be too bad 😄 

>  and then filter out in razor based on context the snippets you want (otherwise these hardcoded ones will be duplicates)

I don't imagine Razor offering any C# built-in snippets out of the box. We'll probably move all this logic over to the C# side in p4 or something.